1단계 : 용어 사전 만들기

MintKim1990 opened the pull request:
안녕하세요. 리뷰어님

1단계 PR 요청드립니다.

용어사전을 만들면서 약간 고민되었던 포인트가 용어들을 만들면서 
비즈니스로직이 포함된 설명을 넣으면서 만들다보니까 
오히려 용어에 대한 정의가 애매모호해지는것같아서 
순수하게 용어에 대한 의미를 설명란에 기재하고

비즈니스로직적인 부분은 2단계 모델링에서 표현하면 어떨까라고 생각하고
작성해봤습니다. 

제가 생각하는 용어사전은 실무에서 PM이나 개발자들이 통상적으로 사용하는 언어를
기재하는것으로 이해하고 엔티티 뿐 아니라 메서드명이나 객체에 상태등을 작성해봤는데

이부분에 대한 리뷰어님 피드백 부탁드립니다 !!
감사합니다 :)

rlaalsxo1990 committed:
feat: 치킨포스 관련한 용어사전 README.md에 작성
 - 치킨포스 도메인을 분석하여 나온 용어들을 용어사전에 작성

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
-| 한글명 | 영문명 | 설명 |
-| --- | --- | --- |
-|  |  |  |
+| 한글명       | 영문명                     | 설명                                        |
```
용어사전(유비쿼터스 언어)은 Bounded Context 내부에서 진정한 힘을 발휘 합니다 😄 
Bounded Context를 기준으로 분리해보는건 어떨까요 ?

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
+| 상품        | Product                 | 판매대상을 의미한다.                               |
+| 상품명       | Product Name            | 상품에 이름을 의미한다.                             |
+| 상품가격      | Product Price           | 상품에 가격을 의미한다.                             |
+| 상품가격변경    | Product Price Change    | 상품에 가격변경을 의미한다.                           |
```
행위를 용어사전에 정의해주셨네요 😄 
개인적으로 행위는 용어사전에 정의하기 보단 모델링에 남기는것이 더 자연스러울것 같은데 어떻게 생각하시나요 ?

MintKim1990 replied:
이부분도 좀 고민되는거같은데

유비쿼터스 언어라는게 객체로 표현할 개념들을 용어로 정리한다면 행위는 제거하는게 맞을듯한데
관계자들과 원활한 커뮤니케이션을 위해 용어를 정리하는 개념이라면 행위도 추가되도 되지않을까?? 
라는 생각에서 저는 후자로 판단하고 추가해보았었는데요.

리뷰어님 생각은 어떠신지 궁금합니다 :)

sah3122 replied:
행위를 추가하는것이 어색하지는 않다고 생각합니다 😄 
다만 일반적으로 사용되는 언어들에 대해서는 용어를 정리하지 않았던것 같아요. 
가령 생성, 수정 등과 같은 용어는 별도로 정리하지 않고 도메인의 핵심 비즈니스 언어를 정의하곤 했습니다 😄 
이번 과정에서는 이야기 주신것처럼 행위를 추가하는것도 좋을것 같아요 🙏

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
+| 착석 테이블    | Occupied OrderTable     | 고객이 착석한 테이블을 의미한다.                        |
+| 비착석 테이블   | Unoccupied OrderTable   | 고객이 착석할 수 있는 테이블을 의미한다.                   |
+| 주문정보      | OrderLineItem           | 주문을 구성하는 메뉴정보를 의미하며 메뉴, 수량을 가진다.          |
+| 주문        | Order                   | 고객이 요청한 주문정보에 집합을 의미한다.                   |
+| 매장주문      | Eatin Order             | 착석테이블 고객의 주문정보에 집합을 의미한다.                 |
+| 포장주문      | Takeout Order           | 포장요청한 고객의 주문정보에 집합을 의미한다.                 |
+| 배달주문      | Delivery Order          | 배달요청한 고객의 주문정보에 집합을 의미한다.                 |
```
프로젝트에 참여하는 이해관계자들이 주문정보의 집합이란 설명을 보고 어떤 내용인지 알수 있을까요 ? 
일반적으로 사용하는 개념이 녹아들면 좋을것 같아요 😄

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
+| 주문요청      | Order Request           | 고객이 주문을 요청하는것을 의미한다.                      |
+| 주문수락      | Order Accept            | 고객이 주문한 요청을 수락하는것을 의미한다.                  |
+| 주문제공      | Order Serve             | 고객이 주문한 메뉴를 제공하는것을 의미한다.                  |
+| 배달시작      | Order Delivery          | 배달주문 요청메뉴 배달을 시작하는것을 의미한다.                |
+| 배달완료      | Order Delivery Complete | 배달주문 요청메뉴 배달이 완료된것을 의미한다.                 |
```
배달 주문에는 배달 대행사, 배달원과 같은 용어도 도출해볼수 있지 않을까요 ?

MintKim1990 replied:
아 그러네요 !!
추가했습니다 :)

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
+| 메뉴그룹      | MenuGroup               | 메뉴들에 집합을 의미하여 이름을 가진다.                    |
+| 테이블       | OrderTable              | 매장내에 고객이 앉을 수 있는 테이블을 의미하여 이름과 고객 수를 가진다. |
+| 착석 테이블    | Occupied OrderTable     | 고객이 착석한 테이블을 의미한다.                        |
+| 비착석 테이블   | Unoccupied OrderTable   | 고객이 착석할 수 있는 테이블을 의미한다.                   |
```
```suggestion
| 비착석 테이블   | Unoccupied OrderTable   | 비어있는 테이블을 의미한다.                   |
```
위와 같이 표현하는것도 가능할것 같아요 😄

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
+| 상품가격변경    | Product Price Change    | 상품에 가격변경을 의미한다.                           |
+| 메뉴상품      | MenuProduct             | 메뉴를 구성하는 상품정보를 의미하며 상품, 수량을 가진다.          |
+| 메뉴상품 금액   | MenuProduct Price       | 메뉴상품에 상품가격 * 수량을 의미한다.                    |
+| 메뉴        | Menu                    | 고객에게 노출되는 메뉴상품에 집합을 의미한다.                 |
```
마찬가지로 `메뉴상품에 집합`이라는 표현 보단 일반적인 설명이 포함되면 좋을것 같아요 😄

sah3122 commented on README.md:
```diff
@@ -96,8 +96,35 @@ docker compose -p kitchenpos up -d
+| 매장주문      | Eatin Order             | 착석테이블 고객의 주문정보에 집합을 의미한다.                 |
+| 포장주문      | Takeout Order           | 포장요청한 고객의 주문정보에 집합을 의미한다.                 |
+| 배달주문      | Delivery Order          | 배달요청한 고객의 주문정보에 집합을 의미한다.                 |
+| 주문요청      | Order Request           | 고객이 주문을 요청하는것을 의미한다.                      |
+| 주문수락      | Order Accept            | 고객이 주문한 요청을 수락하는것을 의미한다.                  |
+| 주문제공      | Order Serve             | 고객이 주문한 메뉴를 제공하는것을 의미한다.                  |
+| 배달시작      | Order Delivery          | 배달주문 요청메뉴 배달을 시작하는것을 의미한다.                |
+| 배달완료      | Order Delivery Complete | 배달주문 요청메뉴 배달이 완료된것을 의미한다.                 |
```
코멘트 남겨주신것처럼 용어 사전에 정의한 영문명은 실제로 코드에 녹아들어야 하는 단어들입니다. 
정의해주신 영문명을 코드로 작성해보았을때 부자연스러운 부분이 없는지 판단해보세요 😄

sah3122 reviewed (changes requested):
용어사전 정리 잘해주셨습니다 👍 
몇가지 코멘트 남겨두었는데 확인 부탁드려요 !

rlaalsxo1990 committed:
fix: 용어사전 Bounded Context 단위로 분리작성
 - 용어를 Bounded Context 단위로 분리하여 작성
 - 행위에 해당하는 용어 제거
 - 용어 내에 관계자들이 이해하기 애매한 용어설명 수정
 - 배달 주문 관련하여 라이더 용어 추가

MintKim1990 commented:
안녕하세요. 리뷰어님 !

피드백 주신사항 수정진행했습니다.
추가적으로 리뷰어님 생각이 궁금한 부분 코멘트 달아놓았는데 확인한번 부탁드립니다.

추가적으로 궁금한 사항이 Bounded Context를 나누는 기준이 좀 고민이 되더라구요 !

예를들어 크게보면 메뉴, 주문으로 나눌수 있을듯하고 
상세하게 나누면 상품, 메뉴, 메뉴그룹, 테이블, 주문으로 나눌 수 있을거같은데

제가생각했을때 Bounded Context를 나누는 기준은 관심사에 따라 나누는게 어떨까라고 생각했습니다.
예를들어

메뉴그룹에 이름이 변경되는게 메뉴가 변경되는건 아니고
상품에 이름이 변경되는게 메뉴가 변경되는건 아니라 생각해서 PR드린 방식으로 나눠봤는데

한편으로는 상품가격이 변경되면 메뉴가 비노출되기도 하니 같은 관심사로 둬야하나?
라는 의문점이 생기더라구요.

리뷰어님은 어떤식으로 Bounded Context를 나누시는지 기준같은게 있을까요?
변경해볼 부분 있으면 재 리뷰요청처리 부탁드립니다 !
감사합니다 :)

sah3122 commented:
바운디드 컨텍스트를 나누는 기준을 명확하게 정의하기는 어려울것 같습니다. 
너무 작은 바운디드 컨텍스트는 분산된 비즈니스 로직들 때문에 복잡도가 올라가고, 반면 너무 큰 바운디드 컨텍스트는 도메인 로직이 비대해지고 복잡해지기 때문에 전체적인 복잡도가 높아지게 됩니다. 

관심사에 따라 나누는것도 하나의 방법이라고 볼수 있을것 같아요. 
저는 Bounded Context를 나눌때 서비스의 단위와 책임을 중요하게 생각합니다. 
상품, 메뉴, 메뉴그룹, 테이블, 주문이라면 상품 서비스, 메뉴 서비스, 주문 서비스로 나뉘어지고 각 서비스들은 다른 서비스와는 관계없이 하나의 책임만을 담당할수 있어야 합니다. 
추가로 컨텍스트의 성장 가능성도 고려하는 편입니다. 특정 컨텍스트의 기능이 핵심 비즈니스를 다루고 있다면 시간이 지날수록 많은 변경과 추가가 일어날것을 예상할수 있기 때문에 별도의 바운디드 컨텍스트로 분리를 하곤 합니다 😄

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
질문에 대한 답변 남겨두었는데 확인 부탁드리며 다음 미션도 화이팅입니다 😄 
미션 진행도중 질문이 있으시면 편하게 DM 주세요 🙏
